File: src/main/java/cz/solutions/cockroach/ExchangeRatesReader.java
package cz.solutions.cockroach;

import java.util.stream.Collectors;
import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Value;
import org.joda.time.LocalDate;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import java.util.Arrays;
import java.util.Map;

public class ExchangeRatesReader {

    public TabularExchangeRateProvider parse(String ... files) {

        Map<LocalDate, Double> mapping = Arrays.stream(files)
                .map(this::parseOne)
                .reduce((firstMap, secondMap) -> {
                    firstMap.putAll(secondMap);
                    return firstMap;
                }).orElse(Map.of());

        return new TabularExchangeRateProvider(mapping);

    }

    public Map<LocalDate, Double> parseOne(String data) {
        String header = data.lines().findFirst().orElseThrow();
        String[] headerParts = header.split("\\|");
        int usdIndex = Arrays.asList(headerParts).indexOf("1 USD");
        return data.lines()
                .skip(1)
                .map(line -> parseLine(line, usdIndex))
                .collect(Collectors.toMap(Line::getDate, Line::getAmount));
    }

    private Line parseLine(String line, int usdIndex) {
        DateTimeFormatter formatter = DateTimeFormat.forPattern("dd.MM.YYYY");
        String[] parts = line.split("\\|");
        return new Line(LocalDate.parse(parts[0], formatter), Money.fromString(parts[usdIndex]));
    }

    @Value
    private static class Line {
        @JsonProperty("Datum")
        @JsonFormat(pattern = "dd.MM.YYYY")
        LocalDate date;
        @JsonProperty("1 USD")
        Money rate;

        public double getAmount() {
            return rate.getAmount();
        }
    }

    @Value
    private static class Money {
        double amount;

        public static Money fromString(String input) {
            return new Money(Double.parseDouble(input.replace(',', '.')));
        }
    }
}

File: src/main/kotlin/cz/solutions/cockroach/ExchangeRatesReader.kt
package cz.solutions.cockroach

import org.joda.time.LocalDate
import org.joda.time.format.DateTimeFormat

object ExchangeRatesReader {

    private val CURRENCY_HEADERS = mapOf(
        Currency.USD to "1 USD",
        Currency.EUR to "1 EUR",
        Currency.GBP to "1 GBP"
    )

    fun parse(vararg files: String): TabularExchangeRateProvider {
        val mapping = files.map { parseOne(it) }
            .reduce { acc, map -> acc + map }
        return TabularExchangeRateProvider(mapping)
    }

    private fun parseOne(data: String): Map<LocalDate, Map<Currency, Double>> {
        val lines = data.lines()
        val header = lines.first()
        val headerParts = header.split("|")
        val indices = CURRENCY_HEADERS.mapValues { (currency, columnHeader) ->
            val idx = headerParts.indexOf(columnHeader)
            require(idx >= 0) { "missing $columnHeader column for $currency in header: $header" }
            idx
        }
        return lines
            .drop(1)
            .filter { it.isNotBlank() }
            .associate { parseLine(it, indices) }
    }

    private fun parseLine(line: String, indices: Map<Currency, Int>): Pair<LocalDate, Map<Currency, Double>> {
        val formatter = DateTimeFormat.forPattern("dd.MM.YYYY")
        val parts = line.split("|")
        val date = LocalDate.parse(parts[0], formatter)
        val rates = indices.mapValues { (_, idx) -> Money.fromString(parts[idx]).amount }
        return date to rates
    }

    private data class Money(val amount: Double) {
        companion object {
            fun fromString(input: String): Money {
                return Money(input.replace(',', '.').toDouble())
            }
        }
    }
}